stdlib/AllocProfile/src/AllocProfile.jl

```diff
@@ -0,0 +1,168 @@
+module AllocProfile
+
+using Base.StackTraces: StackTrace, StackFrame, lookup
+using Base: InterpreterIP
+
+# raw results
+
+# matches RawBacktrace on the C side
+struct RawBacktrace
+    data::Ptr{Csize_t} # in C: *jl_bt_element_t
+    size::Csize_t
+end
+
+# matches RawAlloc on the C side
+struct RawAlloc
+    type::Ptr{Type}
+    backtrace::RawBacktrace
+    size::Csize_t
+end
+
+struct FreeInfo
+    type::Ptr{Type}
+    count::UInt
+end
+
+# matches RawAllocResults on the C side
+struct RawAllocResults
+    allocs::Ptr{RawAlloc}
+    num_allocs::Csize_t
+
+    frees::Ptr{FreeInfo}
+    num_frees::Csize_t
+end
+
+function start(skip_every::Int=0)
+    ccall(:jl_start_alloc_profile, Cvoid, (Cint,), skip_every)
+end
+
+function stop()
+    raw_results = ccall(:jl_stop_alloc_profile, RawAllocResults, ())
+    decoded_results = decode(raw_results)
+    return decoded_results
+end
+
+function clear()
+    ccall(:jl_free_alloc_profile, Cvoid, ())
+end
+
+# top-level entry point
+# TODO: should probably be a macro instead of a
+# higher-order function
+function profile(f::Function, skip_every::Int=0)
+    start(skip_every)
+    res = f()
+    profile = stop()
+    clear()
+    return res, profile
+end
```

Comment:
if we make this a macro, how do we return both the profile and the result of the code that's being profiled? 🤔

Reply:
```
macro profile(ex, skip_every)
   quote
      $start($skip_every)
      local res = $(esc(ex))
      local profile = $stop()
      $clear()
     (res, profile)
   end
end

Reply:
done, thanks :)

is there a way to make `skip_every` be optional, but also come first… 🤔